File: services/worker/scholarsync_worker/citations.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

CitationStyle = Literal["apa", "mla"]


def _display_date(iso_date: str) -> str:
    value = iso_date.replace("Z", "+00:00")
    parsed = datetime.fromisoformat(value)
    return parsed.strftime("%B %-d, %Y")


def format_citation(thread: dict[str, Any], style: CitationStyle) -> str:
    date = _display_date(thread["createdUtc"])
    title = thread["title"].removesuffix(".")
    sentence_title = title if title.endswith((".", "?", "!")) else f"{title}."
    mla_title = title if title.endswith((".", "?", "!")) else f"{title}."

    if style == "mla":
        return f'{thread["author"]}. "{mla_title}" Reddit, r/{thread["subreddit"]}, {date}, {thread["url"]}.'

    return f'{thread["author"]}. ({date}). {sentence_title} Reddit. {thread["url"]}'

File: packages/shared/src/citations.ts
import type { CitationStyle, RedditThread } from "./schemas";

const formatDate = (isoDate: string) => {
  const date = new Date(isoDate);
  return new Intl.DateTimeFormat("en-US", {
    year: "numeric",
    month: "long",
    day: "numeric"
  }).format(date);
};

export function formatCitation(thread: RedditThread, style: CitationStyle) {
  const date = formatDate(thread.createdUtc);
  const title = thread.title.replace(/\.$/, "");
  const sentenceTitle = /[.?!]$/.test(title) ? title : `${title}.`;
  const mlaTitle = /[.?!]$/.test(title) ? title : `${title}.`;

  if (style === "mla") {
    return `${thread.author}. "${mlaTitle}" Reddit, r/${thread.subreddit}, ${date}, ${thread.url}.`;
  }

  return `${thread.author}. (${date}). ${sentenceTitle} Reddit. ${thread.url}`;
}
